### ISSUE TITLE
Single File HTML Report instead of per module

### ISSUE THREAD

**Issue by boris-42** [coveragepy/coveragepy#938]
**Is your feature request related to a problem? Please describe.**

Having whole report in single file has 3 major benefits: 
- Reports may be way more interactive
- It simplifies sharing of them (basically you just share 1 single file vs zip/unzip)
- and the main thing: It will drastically speed up uploading these reports to s3 or file servers (which is done by CI) 

**Describe the solution you'd like**

There are multiple ways to achieve result, as jquery is used probably the simplest one is: 
1) create a lot of blocks (every block is current page of report) 
2) when you open report show the main block with list of modules
3) when you click on any of links hide main block open specific module and update URL address 

This should be quite easy to do and would drastically simplify life (at least my)

**Describe alternatives you've considered**

Alternative we can redo everything in React and add a lot of interactive functionality to report (like search/filtering and so on)

**Additional context**

Is this interesting for someone else? If I spend some time would this get merged?

**Comment by nedbat** [comment]
I would be interested to see this.  Do you have any examples of single-file HTML output that work like this?

**Comment by layoaster** [comment]
I'm also in the situation of a CI pipeline that have the ability to display HTML files (on the pipeline tool UI) so I can see the report main page but since Argo is setting the Content Security Policy [header](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP), CSS styles are disabled and you can't browse the per-module pages.

The pyest-html plugin has an option to create [self contained reports](https://pytest-html.readthedocs.io/en/latest/user_guide.html#creating-a-self-contained-report). @nedbat you can check this sample [single page report [Google Drive]](https://drive.google.com/file/d/1ZKrTbKjFQQ3wAAkNGa_XGLJXPIZ7w5vr/view?usp=share_link)

**Comment by nedbat** [comment]
I'm not opposed to doing something like this, but I don't understand the mechanics.  @layoaster the sample you showed has just a single line of output to show, as an expand/collapse.  The coverage.py reports are more like multi-page reports that have to be re-engineered to be in a single HTML file.

I'm also a bit concerned because some options (like contexts) make for large pages.  Putting all of the pages into a single HTML file would mean one enormous file.  Would it have to be an option? That adds complexity.

**Comment by layoaster** [comment]
Agree with you. The pure-HTML solution for a single page is far from ideal. In my particular case, the size wouldn't be an issue but I understand it might potentially be for most other use cases.

Wish I could do more to help but I have zero skills on the web/frontend side ...

**Comment by bollwyvl** [comment]
Another example of single file tests are the semi-insane [robotframework reports](https://github.com/robotframework/robotframework/tree/master/src/robot/reporting).

These encode the raw data in an efficient way, and then render the HTML from that... but indeed, they are quite large, and not many people can work on them.

Another up-side of single file reports is that they could be more easily embedded inside _other_ reports, such as those generated by pytest or robotframework.

That being said: perhaps it makes the most sense to make this modular, rather than trying to appease everyone with a single approach. e.g. `coverage html --reporter=my_fancy_coverage_reporter:MyReporter` that answers to whatever the internal representation is. This would offer a path forward for #1053.

**Comment by nedbat** [comment]
> That being said: perhaps it makes the most sense to make this modular, rather than trying to appease everyone with a single approach.

We already have the possibility of a modular approach: coverage.py has supported ways to get the data, either through an API, or the JSON report.  Is a separate reporting tool out of the question? Even as a prototyping approach, this seems like a fine path for someone to take.

> This would offer a path forward for https://github.com/nedbat/coveragepy/issues/1053.

That issue is a bit frustrating: I responded within 20 minutes asking for more information, and never heard back from the requester.

**Comment by bollwyvl** [comment]
> the JSON report

Good point: the JSON path sounds the most portable, as it would be less tied to the underlying implementation. Is there a spec for the format, e.g. JSON schema?

> never heard back from the requester

Indeed: requester feedback aside, this is one of those areas where automated tools can get you to _some_ point (e.g. color contrast), but everybody's needs are very different.

**Comment by nedbat** [comment]
> Is there a spec for the format, e.g. JSON schema?

There isn't a spec for the format, but I try hard not to change it.

**Comment by OscarVanL** [comment]
This would be extremely beneficial now as yesterday GitHub Actions added support for non-zipped artifacts uploaded by CI jobs, see [here](https://github.blog/changelog/2026-02-26-github-actions-now-supports-uploading-and-downloading-non-zipped-artifacts/). If you look at the linked blog you can get an idea for what this unlocks.

Here's what I'd like to achieve:
1. Run Python tests
2. Create coverage HTML artifact (single file)
3. Upload as artifact
4. Comment on the PR with link to the HTML report
5. Reviewer can 1 click this link, which will allow them to check the PR has good coverage

Unfortunately this isn't possible with coverage.py, but with my Go tests, the HTML coverage report (via the built-in `go tool cover -html` command) produces a single-file HTML output, so am able to do this.

**Comment by nedbat** [comment]
@OscarVanL thanks, this is useful information.

> Comparably, with my Go tests, the HTML coverage report (via the built-in go tool cover -html command) produces a single-file HTML output, so am able to do this.

Can you give me a link to one of these so I can see how it's made?

**Comment by OscarVanL** [comment]
Hi @nedbat. Thanks for the reply :)

This is the HTML output generated by Go's coverage utility for a small "hello world" example program with partial test coverage:
[coverage.html](https://github.com/user-attachments/files/25608400/coverage.html)

Please check [this repo](https://github.com/OscarVanL/go-coverage-demo) for that demo program. The README says the relevant Go commands to run to generate the coverage report (Note: Repo is AI generated for demo purposes).
